File: backend/src/follower_system.py
"""
Follower system for managing followers in Carcassonne: War of Ages.
"""

import time
import uuid
from typing import Optional, List, Dict
from .models.follower import Follower, FollowerType
from .models.tile import Tile, TileType
from .models.game_state import GameState

RECALL_DURATION = 10.0  # 10 seconds to recall a follower

class FollowerSystem:
    """Manages follower placement, recall, and resource generation."""
    
    @staticmethod
    def can_place_follower(game_state: GameState, player_id: int, tile: Tile, follower_type: FollowerType) -> tuple[bool, str]:
        """Check if a follower can be placed on a tile."""
        player = next((p for p in game_state.players if p.id == player_id), None)
        if not player:
            return False, "Player not found"
            
        if player.followers_available <= 0:
            return False, "No followers available"
            
        if tile.follower_id is not None:
            return False, "Tile already has a follower"
            
        if tile.owner != player_id:
            return False, "You don't own this tile"
            
        # Check follower type compatibility with tile type
        valid_combinations = {
            FollowerType.MAGISTRATE: [TileType.CAPITAL_CITY, TileType.CITY],
            FollowerType.FARMER: [TileType.FIELD],
            FollowerType.MONK: [TileType.MONASTERY],
            FollowerType.SCOUT: [t for t in TileType]  # Scout can claim any tile
        }
        
        if tile.type not in valid_combinations.get(follower_type, []):
            return False, f"{follower_type.value} cannot be placed on {tile.type.value} tiles"
            
        return True, "OK"
    
    @staticmethod
    def place_follower(game_state: GameState, player_id: int, tile_id: str, follower_type: FollowerType) -> Optional[Follower]:
        """Place a follower on a tile."""
        tile = next((t for t in game_state.tiles if t.id == tile_id), None)
        if not tile:
            return None
            
        can_place, _ = FollowerSystem.can_place_follower(game_state, player_id, tile, follower_type)
        if not can_place:
            return None
            
        # Create follower
        follower = Follower(
            id=f"follower_{uuid.uuid4().hex[:8]}",
            player_id=str(player_id),
            type=follower_type,
            tile_id=tile_id
        )
        
        # Update game state
        game_state.followers.append(follower)
        tile.follower_id = follower.id
        
        # Update player's available followers
        player = next(p for p in game_state.players if p.id == player_id)
        player.followers_available -= 1
        
        return follower
    
    @staticmethod
    def start_recall(game_state: GameState, player_id: int, follower_id: str) -> bool:
        """Start recalling a follower."""
        follower = next((f for f in game_state.followers if f.id == follower_id), None)
        if not follower or follower.player_id != str(player_id):
            return False
            
        if follower.is_recalling:
            return False  # Already recalling
            
        follower.is_recalling = True
        follower.recall_started_at = time.time()
        
        return True
    
    @staticmethod
    def complete_recalls(game_state: GameState) -> List[Follower]:
        """Complete any recalls that have finished."""
        completed = []
        current_time = time.time()
        
        for follower in game_state.followers:
            if follower.is_recalling and follower.recall_started_at:
                if current_time - follower.recall_started_at >= RECALL_DURATION:
                    # Remove follower from tile
                    if follower.tile_id:
                        tile = next((t for t in game_state.tiles if t.id == follower.tile_id), None)
                        if tile:
                            tile.follower_id = None
                    
                    # Return follower to player's pool
                    player = next((p for p in game_state.players if p.id == int(follower.player_id)), None)
                    if player:
                        player.followers_available = min(8, player.followers_available + 1)
                    
                    # Remove follower from game
                    game_state.followers.remove(follower)
                    completed.append(follower)
        
        return completed
    
    @staticmethod
    def calculate_resource_generation(game_state: GameState) -> Dict[int, Dict[str, int]]:
        """Calculate resource generation rates for all players."""
        generation_rates = {}
        
        for player in game_state.players:
            if player.is_eliminated:
                continue
                
            rates = {"gold": 0, "food": 0, "faith": 0}
            
            # Get all tiles owned by this player with followers
            player_tiles = [t for t in game_state.tiles if t.owner == player.id and t.follower_id]
            
            # Group connected tiles by type
            connected_groups = FollowerSystem._find_connected_tile_groups(game_state, player.id)
            
            for group in connected_groups:
                # Only need one follower per connected group
                has_follower = any(t.follower_id for t in group)
                if has_follower:
                    tile_type = group[0].type
                    group_size = len(group)
                    
                    # Calculate base generation
                    if tile_type in [TileType.CAPITAL_CITY, TileType.CITY]:
                        # Check if any follower is a magistrate
                        magistrate_present = any(
                            f for f in game_state.followers 
                            if f.tile_id in [t.id for t in group] and f.type == FollowerType.MAGISTRATE
                        )
                        if magistrate_present:
                            rates["gold"] += group_size  # 1 gold per connected city tile
                            
                    elif tile_type == TileType.FIELD:
                        # Check if any follower is a farmer
                        farmer_present = any(
                            f for f in game_state.followers 
                            if f.tile_id in [t.id for t in group] and f.type == FollowerType.FARMER
                        )
                        if farmer_present:
                            rates["food"] += group_size  # 1 food per connected field tile
                            
                    elif tile_type == TileType.MONASTERY:
                        # Check if any follower is a monk
                        monk_present = any(
                            f for f in game_state.followers 
                            if f.tile_id in [t.id for t in group] and f.type == FollowerType.MONK
                        )
                        if monk_present:
                            rates["faith"] += group_size  # 1 faith per connected monastery tile
            
            # Add special resource tiles (mines and orchards) - no follower needed
            for tile in game_state.tiles:
                if tile.owner == player.id:
                    if tile.type == TileType.MINE:
                        rates["gold"] += 2  # 2 gold per mine
                    elif tile.type == TileType.ORCHARD:
                        rates["food"] += 2  # 2 food per orchard
            
            generation_rates[player.id] = rates
        
        return generation_rates
    
    @staticmethod
    def _find_connected_tile_groups(game_state: GameState, player_id: int) -> List[List[Tile]]:
        """Find groups of connected tiles of the same type owned by a player."""
        player_tiles = [t for t in game_state.tiles if t.owner == player_id]
        visited = set()
        groups = []
        
        for tile in player_tiles:
            if tile.id in visited:
                continue
                
            # Start a new group
            group = []
            tile_type = tile.type
            stack = [tile]
            
            while stack:
                current = stack.pop()
                if current.id in visited:
                    continue
                    
                visited.add(current.id)
                group.append(current)
                
                # Check adjacent tiles
                for dx, dy in [(0, 1), (1, 0), (0, -1), (-1, 0)]:
                    nx, ny = current.x + dx, current.y + dy
                    adj_tile = next((t for t in player_tiles if t.x == nx and t.y == ny), None)
                    
                    if adj_tile and adj_tile.id not in visited and adj_tile.type == tile_type:
                        stack.append(adj_tile)
            
            if group:
                groups.append(group)
        
        return groups

File: frontend/src/js/follower-system.js
/**
 * Follower System
 * Manages follower placement, recall, and UI for Carcassonne: War of Ages
 */

export class FollowerSystem {
    constructor(game) {
        this.game = game;
        this.followers = new Map(); // follower_id -> follower data
        this.selectedTileForPlacement = null;
        this.isPlacingFollower = false;
        this.recallingFollowers = new Map(); // follower_id -> recall start time
        
        // Follower types with their descriptions
        this.followerTypes = {
            magistrate: {
                name: "Magistrate",
                description: "Claims cities for gold production",
                icon: "👔",
                validTiles: ["capital_city", "city"]
            },
            farmer: {
                name: "Farmer", 
                description: "Claims fields for food production",
                icon: "🌾",
                validTiles: ["field"]
            },
            monk: {
                name: "Monk",
                description: "Claims monasteries for faith production", 
                icon: "🙏",
                validTiles: ["monastery"]
            },
            scout: {
                name: "Scout",
                description: "Claims any tile for the player",
                icon: "🔍",
                validTiles: null // Can claim any tile
            }
        };
    }
    
    init() {
        this.setupEventListeners();
        this.createFollowerUI();
    }
    
    setupEventListeners() {
        // Listen for follower-related websocket messages
        this.game.websocketClient.on('follower_placed', (data) => this.handleFollowerPlaced(data));
        this.game.websocketClient.on('follower_recalled', (data) => this.handleFollowerRecalled(data));
        this.game.websocketClient.on('follower_recall_complete', (data) => this.handleFollowerRecallComplete(data));
    }
    
    createFollowerUI() {
        // Create follower panel
        const followerPanel = document.createElement('div');
        followerPanel.id = 'follower-panel';
        followerPanel.className = 'follower-panel';
        followerPanel.innerHTML = `
            <h3>Followers</h3>
            <div class="follower-count">
                Available: <span id="followers-available">8</span> / 8
            </div>
            <div class="follower-types">
                ${Object.entries(this.followerTypes).map(([type, info]) => `
                    <button class="follower-type-btn" data-type="${type}" title="${info.description}">
                        <span class="follower-icon">${info.icon}</span>
                        <span class="follower-name">${info.name}</span>
                    </button>
                `).join('')}
            </div>
            <div class="placed-followers">
                <h4>Placed Followers</h4>
                <div id="placed-followers-list"></div>
            </div>
        `;
        
        // Add to UI
        const gameUI = document.querySelector('.game-ui');
        gameUI.appendChild(followerPanel);
        
        // Add event listeners to follower type buttons
        document.querySelectorAll('.follower-type-btn').forEach(btn => {
            btn.addEventListener('click', (e) => {
                const type = e.currentTarget.dataset.type;
                this.startFollowerPlacement(type);
            });
        });
    }
    
    startFollowerPlacement(followerType) {
        if (!this.game.gameState) return;
        
        const player = this.game.gameState.players.find(p => p.id === this.game.playerId);
        if (!player || player.followers_available <= 0) {
            this.game.uiManager.showToast("No followers available!", 'error');
            return;
        }
        
        this.isPlacingFollower = true;
        this.selectedFollowerType = followerType;
        
        // Highlight valid tiles
        this.highlightValidTiles(followerType);
        
        // Update UI
        document.querySelectorAll('.follower-type-btn').forEach(btn => {
            btn.classList.toggle('active', btn.dataset.type === followerType);
        });
        
        this.game.uiManager.showToast(`Select a tile to place ${this.followerTypes[followerType].name}`, 'info');
    }
    
    highlightValidTiles(followerType) {
        if (!this.game.tileSystem || !this.game.gameState) return;
        
        const followerInfo = this.followerTypes[followerType];
        const player = this.game.gameState.players.find(p => p.id === this.game.playerId);
        if (!player) return;
        
        // Clear previous highlights
        this.game.tileSystem.clearHighlights();
        
        // Highlight owned tiles that can accept this follower type
        this.game.gameState.tiles.forEach(tile => {
            if (tile.owner === this.game.playerId && !tile.follower_id) {
                // Check if follower type is valid for this tile
                if (!followerInfo.validTiles || followerInfo.validTiles.includes(tile.type)) {
                    this.game.tileSystem.highlightTile(tile.x, tile.y, 0x00ff00, 0.5);
                }
            }
        });
    }
    
    handleTileClick(x, y) {
        if (!this.isPlacingFollower) return false;
        
        const tile = this.game.gameState.tiles.find(t => t.x === x && t.y === y);
        if (!tile) return false;
        
        // Check if tile is valid for placement
        const followerInfo = this.followerTypes[this.selectedFollowerType];
        if (tile.owner !== this.game.playerId || tile.follower_id) {
            this.game.uiManager.showToast("Cannot place follower on this tile", 'error');
            return true;
        }
        
        if (followerInfo.validTiles && !followerInfo.validTiles.includes(tile.type)) {
            this.game.uiManager.showToast(`${followerInfo.name} cannot be placed on ${tile.type} tiles`, 'error');
            return true;
        }
        
        // Send placement command
        this.placeFollower(tile.id, this.selectedFollowerType);
        
        // Cancel placement mode
        this.cancelFollowerPlacement();
        
        return true; // Handled
    }
    
    placeFollower(tileId, followerType) {
        this.game.websocketClient.send({
            type: 'cmd',
            payload: {
                action: 'placeFollower',
                data: {
                    tile_id: tileId,
                    follower_type: followerType
                }
            }
        });
    }
    
    recallFollower(followerId) {
        this.game.websocketClient.send({
            type: 'cmd',
            payload: {
                action: 'recallFollower',
                data: {
                    follower_id: followerId
                }
            }
        });
    }
    
    cancelFollowerPlacement() {
        this.isPlacingFollower = false;
        this.selectedFollowerType = null;
        this.game.tileSystem.clearHighlights();
        
        document.querySelectorAll('.follower-type-btn').forEach(btn => {
            btn.classList.remove('active');
        });
    }
    
    handleFollowerPlaced(data) {
        const { follower_id, follower_type, tile_id, player_numeric_id, followers_available } = data;
        
        // Add follower to our tracking
        this.followers.set(follower_id, {
            id: follower_id,
            type: follower_type,
            tile_id: tile_id,
            player_id: player_numeric_id
        });
        
        // Update UI if it's our follower
        if (player_numeric_id === this.game.playerId) {
            document.getElementById('followers-available').textContent = followers_available;
            this.updatePlacedFollowersList();
        }
        
        // Update tile visual
        const tile = this.game.gameState.tiles.find(t => t.id === tile_id);
        if (tile && this.game.tileSystem) {
            this.game.tileSystem.addFollowerToTile(tile, follower_type, player_numeric_id);
        }
    }
    
    handleFollowerRecalled(data) {
        const { follower_id, recall_duration } = data;
        
        // Start tracking recall
        this.recallingFollowers.set(follower_id, Date.now());
        
        // Update UI
        this.updatePlacedFollowersList();
        
        // Show visual indicator on tile
        const follower = this.followers.get(follower_id);
        if (follower && this.game.tileSystem) {
            const tile = this.game.gameState.tiles.find(t => t.id === follower.tile_id);
            if (tile) {
                this.game.tileSystem.showRecallAnimation(tile);
            }
        }
    }
    
    handleFollowerRecallComplete(data) {
        const { follower_id, player_id } = data;
        
        // Remove from tracking
        const follower = this.followers.get(follower_id);
        if (follower) {
            // Remove visual from tile
            const tile = this.game.gameState.tiles.find(t => t.id === follower.tile_id);
            if (tile && this.game.tileSystem) {
                this.game.tileSystem.removeFollowerFromTile(tile);
            }
            
            this.followers.delete(follower_id);
        }
        
        this.recallingFollowers.delete(follower_id);
        
        // Update UI
        if (parseInt(player_id) === this.game.playerId) {
            const player = this.game.gameState.players.find(p => p.id === this.game.playerId);
            if (player) {
                document.getElementById('followers-available').textContent = player.followers_available;
            }
            this.updatePlacedFollowersList();
        }
    }
    
    updatePlacedFollowersList() {
        const listContainer = document.getElementById('placed-followers-list');
        listContainer.innerHTML = '';
        
        // Get followers for current player
        const playerFollowers = Array.from(this.followers.values())
            .filter(f => f.player_id === this.game.playerId);
        
        if (playerFollowers.length === 0) {
            listContainer.innerHTML = '<p class="no-followers">No followers placed</p>';
            return;
        }
        
        playerFollowers.forEach(follower => {
            const tile = this.game.gameState.tiles.find(t => t.id === follower.tile_id);
            if (!tile) return;
            
            const followerInfo = this.followerTypes[follower.type];
            const isRecalling = this.recallingFollowers.has(follower.id);
            
            const followerEl = document.createElement('div');
            followerEl.className = 'placed-follower-item' + (isRecalling ? ' recalling' : '');
            followerEl.innerHTML = `
                <span class="follower-info">
                    ${followerInfo.icon} ${followerInfo.name} at (${tile.x}, ${tile.y})
                </span>
                ${!isRecalling ? `
                    <button class="recall-btn" onclick="game.followerSystem.recallFollower('${follower.id}')">
                        Recall
                    </button>
                ` : '<span class="recall-status">Recalling...</span>'}
            `;
            
            listContainer.appendChild(followerEl);
        });
    }
    
    updateGameState(gameState) {
        // Sync followers from game state
        if (gameState.followers) {
            this.followers.clear();
            gameState.followers.forEach(follower => {
                this.followers.set(follower.id, follower);
            });
        }
        
        // Update UI
        const player = gameState.players.find(p => p.id === this.game.playerId);
        if (player) {
            document.getElementById('followers-available').textContent = player.followers_available || 0;
        }
        
        this.updatePlacedFollowersList();
    }
    
    cleanup() {
        this.followers.clear();
        this.recallingFollowers.clear();
        this.cancelFollowerPlacement();
    }
}